### PR TITLE
Simpler configurable emails

### PR DESCRIPTION
**Pull request by pjedrzejewski**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | https://github.com/Sylius/Sylius-Docs/pull/255

Hey folks!

This PR does few things:

1. Makes configuring existing emails so much simpler.
2. Adding new emails is even simpler now.
3. Emails can be configured in the backend (database) and via email.
4. You can send emails via different adapters, we use TwigSwiftmailerAdapter, but you can implement your own and delegate everything to external service/api.

It has some BC breaks, so I am open to discuss when it should be merged.

Things to fix:

* [x] I am sick, so maybe I miss something, but can't get the configuration definition right. (thus the hacks with [1] [0]

Usage:

``` yaml
sylius_mailer:
    sender:
        name: My store
        address: no-reply@mystore.com
    templates:
        ::bar.html.twig
        ::foo.html.twig
    emails:
        foo_email:
            subject: Hello Foo!
            template: foo.html.twig
            enabled: false
```

That would be it. Let me know what do you think. :)

To send the foo_email, you just need to call:

``` php
$this->get('sylius.email_sender')->send('foo_email', array('recipient@example.com'), array('data' => 'bar'));
```

It tries to find the email in DB, then fallbacks to the yaml config. Also, it won't send if it is disabled.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml [hunk 1]
```diff
@@ -246,4 +244,20 @@ sylius_translation:
             fallbackLocale: fallbackLocale
         translation:
             field: translatable
-            locale: locale
\ No newline at end of file
+            locale: locale
+
+sylius_mailer:
```
Comment (amenophis) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: Is driver missing ?

Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: Driver is autoconfigured via sylius_core bundle.

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml [hunk 2]
```diff
@@ -246,4 +244,20 @@ sylius_translation:
...
+    adapter: twig_swiftmailer
+    sender:
+        name: Example.com
+        address: no-reply@example.com
+    emails:
+        order_confirmation:
+            subject: Thank you for your order!
```
Comment (aRn0D) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: Should it  be translatable ?

File: src/Sylius/Bundle/MailerBundle/Sender/Adapter/TwigSwiftMailerAdapter.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    protected $mailer;
+
+    /**
+     * @var Twig_Environment
+     */
+    protected $twig;
+
+    /**
+     * @param DefaultSettingsProviderInterface $defaultSettingsProvider
```
Comment (aRn0D) [inline src/Sylius/Bundle/MailerBundle/Sender/Adapter/TwigSwiftMailerAdapter.php]: alignment ;)

File: src/Sylius/Bundle/MailerBundle/Sender/Adapter/TwigSwiftMailerAdapter.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+     * @param Swift_Mailer $mailer
+     * @param Twig_Environment $twig
+     */
+    public function __construct(
+        DefaultSettingsProviderInterface $defaultSettingsProvider,
+        Swift_Mailer $mailer,
+        Twig_Environment $twig
+    )
+    {
```
Comment (aRn0D) [inline src/Sylius/Bundle/MailerBundle/Sender/Adapter/TwigSwiftMailerAdapter.php]: should go on the previous line

**Comment by amenophis** [comment]
Really nice, will use it standalone as soon as available :)

File: src/Sylius/Component/Mailer/Model/Email.php [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+ */
+
+namespace Sylius\Component\Mailer\Model;
+
+/**
+ * Mailer model.
+ *
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class Email implements EmailInterface
```
Comment (aRn0D) [inline src/Sylius/Component/Mailer/Model/Email.php]: All setter should return $this

File: src/Sylius/Component/Mailer/Provider/EmailProvider.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @var array
+     */
+    protected $configuration;
+
+    /**
+     * @param RepositoryInterface $emailRepository
+     * @param array               $configuration
+     */
+    public function __construct(RepositoryInterface $emailRepository, $configuration)
```
Comment (aRn0D) [inline src/Sylius/Component/Mailer/Provider/EmailProvider.php]: array $configuration

File: src/Sylius/Component/Mailer/Sender/Sender.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Mailer\Sender;
```
Comment (aRn0D) [inline src/Sylius/Component/Mailer/Sender/Sender.php]: Add an extra line

File: src/Sylius/Component/Mailer/Sender/Sender.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+     */
+    protected $adapter;
+
+    /**
+     * @var EmailProviderInterface
+     */
+    protected $provider;
+
+    /**
+     * @param AdapterInterface $adapter
```
Comment (aRn0D) [inline src/Sylius/Component/Mailer/Sender/Sender.php]: alignment

File: src/Sylius/Bundle/CoreBundle/Mailer/Emails.php [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Sylius\Bundle\CoreBundle\Mailer;
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Mailer/Emails.php]: License missing, and new line at end.

File: src/Sylius/Bundle/MailerBundle/Resources/config/doctrine/model/Email.orm.xml [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                  xsi:schemaLocation="http://doctrine-project.org/schemas/orm/doctrine-mapping
+                                      http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd">
+
+    <mapped-superclass name="Sylius\Component\Mailer\Model\Email" table="sylius_email">
+        <id name="id" column="id" type="integer">
+            <generator strategy="AUTO" />
+        </id>
+
+        <field name="code" column="code" type="string" length="255" unique="true" />
```
Comment (stloyd) [inline src/Sylius/Bundle/MailerBundle/Resources/config/doctrine/model/Email.orm.xml]: `length="255"` is default value for `string`. Also adding `column` when you don't change i.e. camelCase is useless.

File: src/Sylius/Component/Mailer/Sender/Adapter/AbstractAdapter.php [hunk 11]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace Sylius\Component\Mailer\Sender\Adapter;
```
Comment (stloyd) [inline src/Sylius/Component/Mailer/Sender/Adapter/AbstractAdapter.php]: Missing license.

File: src/Sylius/Component/Mailer/Sender/Adapter/AdapterInterface.php [hunk 12]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Sylius\Component\Mailer\Sender\Adapter;
```
Comment (stloyd) [inline src/Sylius/Component/Mailer/Sender/Adapter/AdapterInterface.php]: Missing license & new line at end.

File: src/Sylius/Component/Mailer/Sender/SenderInterface.php [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Sylius\Component\Mailer\Sender;
```
Comment (stloyd) [inline src/Sylius/Component/Mailer/Sender/SenderInterface.php]: Missing license & new line at end.

File: src/Sylius/Component/Mailer/composer.json [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        "sylius/resource": "0.13.*@dev"
+    },
+    "require-dev": {
+        "phpspec/phpspec": "~2.1"
+    },
+    "config": {
+        "bin-dir": "bin"
+    },
+    "autoload": {
+        "psr-4": { "Sylius\\Component\\Locale\\": "" }
```
Comment (stloyd) [inline src/Sylius/Component/Mailer/composer.json]: Not `Locale`, but `Mailer`, same below.

**Comment by michalmarcinkowski** [comment]
Generally awesome! :+1:

What I see missing here is ability to add email addresses to CC and BCC.

**Comment by peteward** [comment]
Really good :+1: 
In addition to CC and BCC you might also want attachments (PDF invoices), but it looks pretty easy to extend the relevant adaptor to look for an 'attachments' array within the $data, and it would probably be something more on a per-client basis anyway, definitely not worth stopping this as it is.

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml (hunk 1 above)
Comment (amenophis) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: OK ;)

**Pull request merged**